File: app/tracing.py
from __future__ import annotations

import logging

logger = logging.getLogger(__name__)

_tracing_ready = False


def setup_tracing(service_name: str = "comfy-meta-viewer", level: str = "INFO") -> None:
    global _tracing_ready
    if _tracing_ready:
        return

    try:
        from opentelemetry import trace
        from opentelemetry.sdk.resources import SERVICE_NAME, Resource
        from opentelemetry.sdk.trace import TracerProvider
        from opentelemetry.sdk.trace.export import BatchSpanProcessor, ConsoleSpanExporter
        from opentelemetry.sdk.trace.sampling import ALWAYS_ON
    except ImportError:
        logger.debug("opentelemetry not installed; tracing disabled")
        return

    resource = Resource.create({SERVICE_NAME: service_name})
    provider = TracerProvider(resource=resource, sampler=ALWAYS_ON)

    import os
    if os.environ.get("CMV_CONSOLE_TRACING") == "1" or os.environ.get("OTEL_TRACES_EXPORTER") == "console":
        provider.add_span_processor(BatchSpanProcessor(ConsoleSpanExporter()))
        logger.info("opentelemetry tracing enabled -> stdout (console exporter)")

    trace.set_tracer_provider(provider)
    _tracing_ready = True


def get_tracer(name: str = __name__):
    try:
        from opentelemetry import trace

        return trace.get_tracer(name)
    except ImportError:
        return _NoOpTracer()


class _NoOpTracer:
    def start_as_current_span(self, *args, **kwargs):
        import contextlib

        return contextlib.nullcontext()

    def start_span(self, *args, **kwargs):
        return None

File: app/static/js/tracing.js
let _ready = false;
let _tracer = null;
const _startTime = typeof performance !== "undefined" ? performance.now() : Date.now();

export function now() {
    return (typeof performance !== "undefined" ? performance.now() : Date.now()) - _startTime;
}

export async function setupTracing() {
    if (_ready || typeof window === "undefined") return;

    try {
        const otelApi = await import("@opentelemetry/api");
        const { WebTracerProvider } = await import("@opentelemetry/sdk-trace-web");
        const { Resource } = await import("@opentelemetry/resources");
        const { SemanticResourceAttributes } = await import("@opentelemetry/semantic-conventions");
        const { SimpleSpanProcessor, ConsoleSpanExporter } = await import("@opentelemetry/sdk-trace-base");
        const { XMLHttpRequestInstrumentation } = await import("@opentelemetry/instrumentation-xml-http-request");

        const provider = new WebTracerProvider({
            resource: new Resource({
                [SemanticResourceAttributes.SERVICE_NAME]: "comfy-meta-viewer-web",
            }),
        });
        provider.addSpanProcessor(new SimpleSpanProcessor(new ConsoleSpanExporter()));

        provider.register();

        try {
            const xhrInst = new XMLHttpRequestInstrumentation({
                ignoreUrls: [/esm\.sh/],
            });
            xhrInst.setTracerProvider(provider);
            xhrInst.enable();
        } catch (e) {
            console.warn("[tracing] xhr instrumentation failed", e);
        }

        _tracer = otelApi.trace.getTracer("comfy-meta-viewer-web");
        _ready = true;
        console.info("[tracing] opentelemetry web tracing enabled -> console");
    } catch (e) {
        console.warn("[tracing] setup failed (cdn/importmap issue)", e);
        _tracer = _noopTracer();
        _ready = true;
    }
}

function _noopTracer() {
    return {
        startSpan(name) {
            return _noopSpan(name);
        },
        startActiveSpan(name, fn) {
            return fn(_noopSpan(name));
        },
    };
}

function _noopSpan(name) {
    const ts = now();
    return {
        setAttribute() {},
        setAttributes() {},
        addEvent() {},
        end() {},
        spanContext() {
            return { traceId: "noop", spanId: "noop", traceFlags: 0 };
        },
        _name: name,
        _ts: ts,
    };
}

export function getTracer() {
    if (!_tracer) {
        _tracer = _noopTracer();
    }
    return _tracer;
}

export function traceSpan(name, attrs = {}) {
    const t = getTracer();
    return t.startSpan(name, { attributes: attrs });
}

export async function traceAsync(name, attrs = {}, fn) {
    const t = getTracer();
    const span = t.startSpan(name, { attributes: attrs });
    try {
        const result = await fn(span);
        span.end();
        return result;
    } catch (err) {
        span.setAttribute("error", true);
        span.setAttribute("error.message", err?.message || String(err));
        span.end();
        throw err;
    }
}
